lib/task/migrate/arUpgradeSqlTask.class.php

```diff
@@ -149,88 +150,117 @@ protected function execute($arguments = array(), $options = array())
     // Find all the upgrade classes in lib/task/migrate
     $version = $this->initialVersion;
 
-    foreach (sfFinder::type('file')
-              ->maxdepth(0)
-              ->sort_by_name()
-              ->name('arUpgrader*.class.php')
-              ->in(sfConfig::get('sf_lib_dir').'/task/migrate') as $filename)
+    if (!empty($options['number']))
     {
-      $className = preg_replace('/.*(arUpgrader\d+).*/', '$1', $filename);
-      $class = new $className;
-
-      if ($class::INIT_VERSION > $version)
-      {
-        continue;
-      }
-
-      try
+      foreach (explode(',', $options['number']) as $number)
       {
-        if ($options['verbose'])
+        if (is_numeric($number))
         {
-          $this->logSection('upgrade-sql', sprintf('Upgrading from Release %s', $class::MILESTONE));
-        }
+          $className = sprintf('arMigration%04d', $number);
 
-        while ($class->up($version, $this->configuration, $options))
-        {
-          // Update version in database
-          $this->updateDatabaseVersion(++$version);
+          if (!class_exists($className))
+          {
+            $this->logSection('upgrade-sql', sprintf('Migration %d not found', $number));
+
+            continue;
+          }
+
+          $this->logSection('upgrade-sql', sprintf('Running migration %d', $number));
+
+          $class = new $className;
+
+          if (true !== $class->up($this->configuration))
+          {
+            throw new sfException(sprintf('Failed to apply upgrade %s', get_class($class)));
+          }
         }
       }
-      catch (Exception $e)
+    }
+    else
+    {
+      foreach (sfFinder::type('file')
```

Comment:
I think it would be better to factor this (main) execution branch into a separate method to aid code readability.